### DISCUSSION TITLE
Tailwindcss and CSS Modules

### DISCUSSION THREAD

**Discussion by jkupcho** [vercel/next.js#18721] [Help]
I'm basing off of the existing sample project:
https://github.com/vercel/next.js/tree/canary/examples/with-tailwindcss

I can't seem to get tailwind CSS `@apply` to work with CSS Modules. Is there a way to do this, or is the intention that all styling using tailwind would take place within `styles/index.css`

Thanks!

**Comment by agcty** [comment]
Hey! I always had the same problem but the solution is actually much easier than you think: just don't use css modules at all.
 
Tailwind was built with a utility-first approach in mind, that means you use classes like "mr-12 border bg-red-500" instead of using custom css and you are also free to extend the tailwind config as needed to suit your project. This pretty much accounts for 99.5% of what you need.

If I really need some custom css which I rarely do, I just use styled jsx inside the component itself which makes it incredibly easy to read. Scattering your css across modules makes your code harder to read & maintain IMO. 

When building a tailwind-styled component I do it like this:

1. come up with the structured html (so basically only the semantics, don't add classes just yet)

 ```html
<div>
   <h1>Title</h1>
   <p>This is a small paragraph</p>
   <img
      src="https://images.unsplash.com/photo-1604321951949-e9d5710716db?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=3578&q=80"
      width="400"
      height="400"
      />
</div>
```

2. Add tailwind classes

```html
<div className="p-8 border border-gray-600 rounded-lg">
   <h1 className="text-2xl font-semibold">Title</h1>
   <p className="mt-1 font-semibold">This is a small paragraph</p>
   <img
      src="https://images.unsplash.com/photo-1604321951949-e9d5710716db?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=3578&q=80"
      width="400"
      height="400"
      className="mt-6 shadow-lg"
   />
</div>
```


3. If you need custom css just use styled-jsx which is included in next.js by default

```html
<div className="p-8 border border-gray-600 rounded-lg">
   <h1 className="text-2xl font-semibold">Title</h1>
   <p className="mt-1 font-semibold">This is a small paragraph</p>
   <img
      src="https://images.unsplash.com/photo-1604321951949-e9d5710716db?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=3578&q=80"
      width="400"
      height="400"
      className="mt-6 shadow-lg custom-blur"
      />


   <style jsx>
      {`
       .custom-blur {
         filter: blur(20px) saturate(180%);
        }
      `}
   </style>
</div>
```

Here's a sample of all three so you can see the visual difference:

![image](https://user-images.githubusercontent.com/12101091/98158741-b9def580-1edb-11eb-8901-5e80d72d53cc.png)

That being said, if you still need global css, just use the global tailwind.css file which you import into your _app component.

**Reply by jkupcho** [reply]
Thanks for replying! I'll have to give this a go!

**Reply by Jbmanllr** [reply]
Hello, im buliding a site based on the next js commerce template, in this template they actually use tailwind along with css module, so you could check how they did. I dont want to do like @agcty suggested because i would like my components to be framework agnostic, and be able to fully change a theme with one click. But i'm still wondering if i should use CSS MODULE to do that or there is a better solution?

**Comment by sumanthratna** [comment]
just want to add: I think you can use CSS modules with tailwind—just rename your `myfile.module.css` to `myfile.module.scss`. I haven't tested if this actually addresses your problem, but SCSS modules work perfectly fine for me!
